### DISCUSSION TITLE
Stripe SDK + Edge Function Compatibility

### DISCUSSION THREAD

**Discussion by treyhuffine** [vercel/community#1124] [Idea]
Stripe updated their SDK so it would work in V8 environments by allowing for the `fetch` API instead of `net/http`. However, it doesn't work in Vercel edge functions because of the Node events library - https://github.com/stripe/stripe-node/blob/dab331e25c8aa1a0c78faff0c4f862f5eb8567b8/src/stripe.ts#L48

It shows the error `Error: The edge runtime does not support Node.js 'events' module.`

I imagine this is one of the most-used packages, and I'm hopeful there's a solution because Cloudflare was able to use the Stripe SDK in their Workers' V8 environment - https://blog.cloudflare.com/announcing-stripe-support-in-workers/

**Comment by amyegan** [comment]
Thanks for reaching out @treyhuffine! One of the [limitations of Edge Functions](limitations) is they do not support native Node.js APIs. So [Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions#) might be a better option in this case.

I'm going to ask the team to see if anyone knows a way to get the Stripe SDK working in Edge Functions, and I'll make sure it's on the list of feature requests if there isn't already a way to make it work.

**Comment by treyhuffine** [comment]
Thank you for the quick response @amyegan! Since the architecture for Cloudflare Workers is the same as Vercel Edge Functions, they also have the same restriction of not allowing for Node.js modules, just like Vercel. I know Cloudflare collaborated directly with Stripe to make updates to the Stripe SDK so it would work in the Cloudflare V8 environment, which sounds very similar to the architecture Vercel uses for Edge Function with V8.

From the [Cloudflare article](https://blog.cloudflare.com/announcing-stripe-support-in-workers/):

<img width="788" alt="Screen Shot 2022-12-12 at 5 08 58 PM" src="https://user-images.githubusercontent.com/11709986/207376635-f0f57a80-eea4-4a84-b3c5-965f9f789752.png">

For example, to enable this to work in V8, Stripe introduced the ability in their SDK to inject a `fetch` function so it could pull it from the V8 engine like the browser does and not rely on the Node `http` module. 

```javascript
const Stripe = require("stripe");

const stripe = Stripe(STRIPE_API_KEY, {
  // This uses a native `fetch` now so it can work outside Node
  httpClient: Stripe.createFetchHttpClient()
});
```

[Stripe also added try/catch for native Node requirements](https://github.com/stripe/stripe-node/blob/3e80143129af5046c42cbec338c9049aa525face/src/utils.ts) so it can fail non-essential aspects gracefully since it's not available in the Cloudflare/Vercel V8 architecture.

<img width="613" alt="Screen Shot 2022-12-12 at 5 17 41 PM" src="https://user-images.githubusercontent.com/11709986/207376717-aaf665c1-5f5d-48ad-88e1-82d4d9310619.png">

With payments being such a core part of applications and a majority of apps using Stripe for it, it would be great to cover this and utilize it through Edge Functions since they seem to be much faster than Serverless Functions which require a cold start. The improved speed of Edge Function would provide a better user experience at a key piece of the user’s journey.

Perhaps Vercel has a contact with the Stripe team? I believe they have committed code to Next.js for Stripe implementations.

What can I do to help move this forward?

**Reply by amyegan** [reply]
Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can share with the team to illustrate how you have set this up on your site?

**Reply by amyegan** [reply]
After a little digging, it looks like Stripe is not entirely able to run in Cloudflare either due to some Node API dependencies (https://github.com/stripe/stripe-node/issues/1593). A non-Node version, as mentioned in https://github.com/stripe/stripe-node/issues/1593#issuecomment-1335851167, would be the neatest option to make the Stripe SDK fully compatible with all edge function platforms (not just Vercel or Cloudflare). 🙏 

The Vercel Edge team has recently been focused on preparation for today's [Edge Functions GA release](https://vercel.com/blog/edge-functions-generally-available). But I expect they will be adding more features to address Node.js compatibility and other concerns in the near future.

I think the simplest solution is to use a Serverless Function for now, and convert it into an Edge Function once the Stripe SDK can be fully supported. The guide on [improving lambda performance](https://vercel.com/guides/how-can-i-improve-serverless-function-lambda-cold-start-performance-on-vercel#improving-your-lambda-performance) has some tips that can reduce the potential cold start impact. 🙇

**Reply by treyhuffine** [reply]
Thank you for digging @amyegan! It looks like that issue you shared from Cloudflare was actually mislabled by the submitter, and the developer got the Pages Functions confused with Workers where the Node Stripe SDK does run. I can confirm that the Stripe SDK does work in the Cloudflare V8 environment, but it does look like they maybe had to stub some functions that use Node APIs that aren't actually critical to running the SDK, or there was a browser equivalent that could be substituted in.

I understand it's challenging to prioritize everything, especially when a feature is new and moving quickly, but I think with Stripe being the primary payments platform used by devs and the effort that the Stripe team put in to make the SDK's core functionality work in V8 environments that value to Vercel users would be high and hopefully the effort from your team would be reasonable.

If there is anything that I can do to help push this forward, let me know!

**Reply by amyegan** [reply]
In terms of what you can do to push this forward, you have already done the most important thing! Hearing about these pain points really helps direct design and development effort where it is most needed. 🙇 Thanks for sharing your use case and for being so understanding about development time and prioritization!

The good news is, Edge team is already planning this work. 🙂 They're looking at the requirements of Stripe SDK, and many other popular tools, to make sure Edge Functions have the necessary pieces to support existing Node tools. I don't have an ETA yet, but I know the team is looking at this very closely to design a thoughtful, sustainable solution.

**Reply by znwhite** [reply]
I would like to comment quickly and say that I had a website stop communicating with Stripe API (requests were stalling without error responses) and setting my Stripe config to include `httpClient: Stripe.createFetchHttpClient()` fixed the issue across all functions. The site's whole API are Vercel functions that live in the `pages/api` folder.

The issue was only happening in production, not locally.

I have no idea why this became an issue suddenly this morning, but thankfully I found this post and doing what I mentioned above ended up fixing the issue. Pretty scary and frustrating, but thank you for writing this up and I hope anyone experiencing the same issue finds this comment and gives it a try.

**Reply by snayerman** [reply]
@znwhite we also ran into this issue today, we've created tickets with vercel as this was blocking our deployments today. About to run a test build with your fix, fingers crossed and will update 🤞 

UPDATE: IT WORKS. Thank you so much @znwhite!!!!!!!!!

**Comment by treyhuffine** [comment]
Thanks @amyegan that's very exciting to hear

**Comment by stephenasuncionDEV** [comment]
Just wondering if this is still being worked on

**Reply by amyegan** [reply]
The team is working on expanding Edge Function capabilities, but I don't have an ETA for additions that would enable Stripe SDK to run in Edge Functions

**Comment by bijanv** [comment]
I'm still not able to resolve this problem on my end. 

- I've confirmed that my api route is deploying to a serverless function and not on an edge function (I'm checking my vercel usage dashboard and edge function requests are not getting any requests). 
- I've added the httpClient  
`const stripe = new Stripe(process.env.STRIPE_SECRET_KEY, {
    apiVersion: '2022-11-15',
    httpClient: Stripe.createFetchHttpClient(), 
  });`
- I'm still not able to get request to Stripe to stop stalling as @znwhite has described on production

Any other ideas on how to debug and address this issue?

**Reply by rajbala** [reply]
@bijanv Did you ever find a resolution to this problem?  I also have tried setting `httpClient: Stripe.createFetchHttpClient()` but it does not change the outcome.  Calls to Stripe just silently fail.

**Comment by z-x** [comment]
I faced the same issue in my SvelteKit project, so if this would be possible to solve, it would be great.

My current solution was to remove the stripe-related code from the edge, and since it is super easy in SvelteKit (per route) it seems to do the trick for now.

**Comment by treyhuffine** [comment]
@znwhite Are you running this on the edge or serverless?

**Reply by mchennupati** [reply]
I guess this hasn't been addressed yet.  On the fly checking if a user is still a subscriber when running LLMs etc on the edge environment would not be possible right ? Or am i missing something. I get this error, when i try to isolate the stripe call, but vercel clearly doesn't like it :)


./node_modules/function-bind/implementation.js
--
15:51:13.376 | Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime
15:51:13.376 | Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation
15:51:13.376 |  
15:51:13.376 | Import trace for requested module:
15:51:13.376 | ./node_modules/function-bind/implementation.js
15:51:13.376 | ./node_modules/function-bind/index.js
15:51:13.376 | ./node_modules/call-bind/index.js
15:51:13.376 | ./node_modules/call-bind/callBound.js
15:51:13.377 | ./node_modules/side-channel/index.js
15:51:13.377 | ./node_modules/qs/lib/stringify.js
15:51:13.377 | ./node_modules/qs/lib/index.js
15:51:13.377 | ./node_modules/stripe/lib/utils.js
15:51:13.377 | ./node_modules/stripe/lib/stripe.js
15:51:13.377 | ./utils/isUserSubscriber.js
15:51:13.377 |  
15:51:13.378 |  
15:51:13.378 | > Build failed because of webpack errors

**Reply by treyhuffine** [reply]
It doesn't seem to be addressed. I'm pretty sure Vercel edge runs on Cloudflare too which explicitly enabled Stripe.

**Reply by mchennupati** [reply]
Something like this works for us.

export const isUserSubscriber = async (userId) => {
  try {
    const customerId = await getCustomerIdByUserId(userId);
    if (customerId === "nostripeidfound") {
      throw new Error("No Stripe customer ID found for user");
    }
    let isActive = false;

    const response = await fetch(
      `https://api.stripe.com/v1/subscriptions?customer=${customerId}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${process.env.STRIPE_SECRET_KEY}`,
        },
      }
    );
    const subscriptions = await response.json();
    console.log("Subscriptions:", subscriptions);

    isActive = subscriptions.data.some(
      (subscription) => subscription.status === "active"
    );

    return isActive;
  } catch (error) {
    console.error("Error fetching subscriptions:", error);
    return false;
  }
};

**Comment by ctjlewis** [comment]
I wrote [`stripe-edge`](https://www.npmjs.com/package/stripe-edge) a few months ago - not typed though.

**Comment by venkatd** [comment]
For those of you who don't use the stripe-sdk heavily, we internally use a much more lightweight HTTP client.

We still get a good amount of type safety by leaning on the types from the stripe sdk but without importing any actual code.

If you take a look at the curl requests in Stripe's documentation, you'll know what the paths are. If you're using Node 18+, cloudflare workers, basically anything environment with fetch, this "SDK" should do the trick.

You may need to flesh it out further--we were minimalistic for our own internal needs.

Hope this helps!

Here's the code below:

```typescript
import type { Stripe } from 'stripe' // only import types for adding safety to the requests

const STRIPE_API_URL = 'https://api.stripe.com/v1'

export class StripeHTTPClient {
  constructor(private secretKey: string) {}

  async get<T>(path: string) {
    return this.request<T>('GET', path)
  }

  async put<T>(path: string, body: Record<string, any>) {
    return this.request<T>('PUT', path, body)
  }

  async post<T = unknown>(path: string, body: Record<string, any>) {
    return this.request<T>('POST', path, body)
  }

  async request<T = unknown>(
    method: 'GET' | 'POST' | 'PUT',
    path: string,
    body?: Record<string, any>
  ) {
    const headers = {
      'Authorization': `Bearer ${this.secretKey}`,
      'Content-Type': 'application/x-www-form-urlencoded'
    }

    const encodedBody = body ? encodeFormUrlEncoded(body) : undefined

    const response = await fetch(`${STRIPE_API_URL}${path}`, {
      method,
      headers,
      body: encodedBody
    })

    if (!response.ok) {
      const text = await response.text()
      throw new Error(`Stripe API error: ${text}`)
    }

    return response.json().then((data) => data as Stripe.Response<T>)
  }
}

function encodeFormUrlEncoded(
  data: Record<string, any>,
  prefix?: string
): string {
  const params: string[] = []

  for (const key in data) {
    if (data.hasOwnProperty(key)) {
      const value = data[key]
      if (value === null || value == undefined) continue

      let finalKey = prefix ? `${prefix}[${key}]` : key

      if (typeof value === 'object') {
        params.push(encodeFormUrlEncoded(value, finalKey))
      } else {
        params.push(`${finalKey}=${encodeURIComponent(value)}`)
      }
    }
  }

  return params.join('&')
}
```

Here's some sample usage:
```typescript
/**
 * Create a payment session for a subscription.
 * Using this session, a Stripe checkout interface can take a user
 * through a checkout flow and, ultimately, create a recurring subscription for the user.
 *
 * @param
 * @returns
 */
export async function createOrganizationSubscriptionPaymentSession({
  stripePriceId,
  organizationId,
  billingEmail,
  trialPeriodDays,
  successRedirect,
  failureRedirect
}: Opts): Promise<Stripe.Response<Stripe.Checkout.Session>> {
  const stripe = new StripeHTTPClient(process.env.STRIPE_SECRET_KEY!)

  if (trialPeriodDays && trialPeriodDays > MAX_TRIAL_PERIOD_DAYS) {
    throw `Max trial period days is ${MAX_TRIAL_PERIOD_DAYS}`
  }

  const appUrl = process.env.NEXT_PUBLIC_APP_URL

  const metadata = { organizationId }

  const sessionData: Stripe.Checkout.SessionCreateParams = {
    metadata: metadata,
    customer_email: billingEmail,
    client_reference_id: organizationId,
    subscription_data: {
      metadata: metadata,
      trial_period_days: trialPeriodDays
    },
    line_items: [
      {
        // Provide the exact Price ID (for example, pr_1234) of the product you want to sell
        price: stripePriceId,
        quantity: 1
      }
    ],
    mode: 'subscription',
    allow_promotion_codes: true,
    success_url: `${appUrl!}/api/payment-success/{CHECKOUT_SESSION_ID}?redirect=${encodeURIComponent(
      successRedirect
    )}`,
    cancel_url: `${appUrl!}/${failureRedirect}`
  }

  return stripe.post<Stripe.Checkout.Session>('/checkout/sessions', sessionData)
}
```
